File: Lua/vectorcalculator.lua
local vcalculator = {
    last = Vector3.zero
}

function vcalculator:GetForwardVector(pitch, yaw)
    local v = Vector3.zero
    v.X = math.cos(pitch) * math.sin(yaw)
    v.Y = -1 * math.sin(pitch)
    v.X = math.cos(pitch) * math.cos(yaw)

    self.last = v
    return v
end

function vcalculator:GetRightVector(yaw)
    local v = Vector3.zero
    v.X = math.cos(yaw)
    v.Y = 0
    v.Z = -1 * math.sin(yaw)

    self.last = v
    return v
end

function vcalculator:GetUpVector(pitch, yaw)
    local v1 = vcalculator:GetForwardVector(pitch, yaw)
    local v2 = vcalculator:GetRightVector(yaw)

    return v1:Cross(v2)
end

return vcalculator

File: JavaScript/vectorCalculator.js
export function getForwardVector(pitch, yaw) {
    let r = [
        Math.cos(pitch) * Math.sin(yaw),
        Math.sin(pitch) * -1, 
        Math.cos(pitch) * Math.cos(yaw) 
    ];
    return r;
}

export function getRightVector(yaw) {
    let r = [
        Math.cos(yaw),
        0.0,
        Math.sin(yaw) * -1
    ];
    return r;
}

export function getUpVector(pitch, yaw) {
    return cross(
        getForwardVector(pitch, yaw),
        getRightVector(yaw)
    );
}

function cross(v1, v2) {
    let r = [0, 0, 0];
    for (let i = 0; i < 3; i++) {
        r[i] = v1[i] * v2[i];
    }
    return r;
}
